cfpb/consumerfinance.gov
# Trashless Wagtail

We disabled Wagtail's built-in deletion back in 2016, for good reason at the time. In its place we created a "Trash" site in Wagtail and our approach was to move content to that site rather than delete it. 

This PR does four things:

- Restores the ability to delete pages in Wagtail, in https://github.com/cfpb/consumerfinance.gov/pull/8310/commits/6ba1b74aaea397435b37e9d7a3778a912d1cbae2. 
   
   We had disabled deletion in `before_delete_page` and `before_bulk_action` Wagtail hooks, and this commit removes those.

- Prevents deletion of locked pages in Wagtail, in https://github.com/cfpb/consumerfinance.gov/pull/8310/commits/87f6ec69d5b86d8d7523f701b9bbcaa7ac96a7ba. 

   Bizarrely, locked pages are still deletable, an issue that's been open in Wagtail [since 2015](https://github.com/wagtail/wagtail/issues/1957). This commit adds in `before_delete_page` and `before_bulk_action` Wagtail hooks that check whether a page is locked before proceeding with the deletion.

- Exports a JSON dump of page data (limited to the data directly belonging to a page, not related data) to a configured location before deletion and enables importing of that data via the admin, in https://github.com/cfpb/consumerfinance.gov/pull/8310/commits/6f0fe69bb7f148a1307d837ebed15f5b5aaa55d9.

   On the off chance that a page is accidentally deleted, want to be able to recover that page with the ease that restoring the entire database from backup is not. 

   This commit uses the `before_delete_page` and `before_bulk_action` Wagtail hooks to connect a [Django `pre_delete` signal handler to the `Page` model](https://github.com/cfpb/consumerfinance.gov/pull/8310/commits/6f0fe69bb7f148a1307d837ebed15f5b5aaa55d9#diff-31e54f1eec8fdc5c71e9b0e4bc91ac3a03c7afcff6954845873a6165c897fdd8R20-R62) that will [export the `specific` page's `serializable_data`](https://github.com/cfpb/consumerfinance.gov/pull/8310/commits/6f0fe69bb7f148a1307d837ebed15f5b5aaa55d9#diff-fbe9b1e780892a6d6ed4c3bde91e26ef435bbac5f68f0221d95820880fdc8dc4R16-R62). This JSON is intended to be reasonably human-readable so that an editor could reconstruct the page from it if needed.

   This commit also adds an `Import` button to Wagtail via [standard hooks](https://github.com/cfpb/consumerfinance.gov/pull/8310/commits/6f0fe69bb7f148a1307d837ebed15f5b5aaa55d9#diff-31e54f1eec8fdc5c71e9b0e4bc91ac3a03c7afcff6954845873a6165c897fdd8R90-R105) that consists of [a view](https://github.com/cfpb/consumerfinance.gov/pull/8310/commits/6f0fe69bb7f148a1307d837ebed15f5b5aaa55d9#diff-60891f147bea2ba2982ed7c6100bc9083a6ec64a65b52c23499b9ffebfa7cc98R25-R47), [template](https://github.com/cfpb/consumerfinance.gov/pull/8310/commits/6f0fe69bb7f148a1307d837ebed15f5b5aaa55d9#diff-9965ef79ea4e60ef6e8e23ade14fadddff021d2780e7148220261ff36b9dfe1e), and [import function](https://github.com/cfpb/consumerfinance.gov/pull/8310/commits/6f0fe69bb7f148a1307d837ebed15f5b5aaa55d9#diff-fbe9b1e780892a6d6ed4c3bde91e26ef435bbac5f68f0221d95820880fdc8dc4R65-R118) that will create a new page from exported page data with the click of a couple of buttons.

   There are significant limitations to the data that gets exported and imported: our recently added Wagtail footnotes is an example. Footnotes are stored as references in a separate table *to* the page. Because they're not stored on the page itself, they will not be exported, but the references on the page will be. This will break on import without editing the JSON data.

   The JSON export and import is *hypothesized* to be useful. ***If we find we are not using it, and that deletion is happening without the need to restore content with any significant frequency, we should revisit and remove this functionality.***

- Modifies the `EmailSignUp` snippet's privacy disclosure page field to `SET_NULL` when the destination page is deleted, in https://github.com/cfpb/consumerfinance.gov/pull/8310/commits/189314f5538e6b202e7f2e325a14dd2261906072. This means if we delete a disclosure page that an email signup is using, it will remove that page from the email signup. An editor will need to set a new one.


## How to test this PR

There are a few things to test:

Deletion and JSON export

1. Set `ARCHIVE_FILESYSTEM` in the environment, i.e. `export ARCHIVE_FILESYSTEM=/tmp/json_archives`. Run cf.gov.
2. Find a page and delete it, observe that deletion occurs.
3. Find a page with child pages and delete it, observe that deletion occurs for all child pages.
4. Select multiple pages and delete them, observe that deletion occurs.
5. Visit the directory set in `ARCHIVE_FILESYSTEM` and observe that there are JSON files for all pages deleted, in their site slug paths.

Import a deleted page
1. Choose a JSON file deleted above, and use the `Import` action in the Wagtail page explorer "…" menu to import it.

Locked pages
1. Lock a page and attempt to delete it. Observe that deletion cannot occur.

## Checklist
